### PR TITLE
Support TS 4.7 moduleResolution Node16/NodeNext

### LINKED ISSUE #575: Support projects with TypeScript 4.7 compiler option "moduleResolution" "Node16"/"NodeNext"
For projects that uses the new `moduleResolution` compiler option value `Node16` or `NodeNext`, TypeScript types for all dependency packages also needs to be exported through the exports section of package.json.

I will create a PR to mitigate this.

Reference: [Announcing TypeScript 4.7](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing)

### PR DESCRIPTION
**Pull request by josundt**
Closes/fixes #575 

### Checklist

- [ ] This PR makes changes to the public API 
- [ ] I have included links for closing relevant issue numbers

### CODE REVIEW

**Comment by pamapa** [comment]
does that work with pre Typescript 4.7 as well? E.g. 4.6 and 4.5?

**Review by kherock — COMMENTED** [review]
@pamapa TypeScript does not attempt to evaluate export maps without enabling one of the new `Node*` moduleResolution options, so this should not impact compatibility with older TS versions.

File: package.json [hunk 1]
```diff
@@ -12,8 +12,14 @@
   "types": "dist/types/oidc-client-ts.d.ts",
   "exports": {
     ".": {
-      "import": "./dist/esm/oidc-client-ts.js",
-      "require": "./dist/umd/oidc-client-ts.js"
+      "import": {
+        "types": "./dist/types/oidc-client-ts.d.ts",
```
[the remark this answers is not in the document]
Comment (josundt) [reply package.json]: @kherock You're right. The export syntax could be simplified, and I have updated the PR accordingly.

File: package.json [hunk 2]
```diff
@@ -12,14 +12,9 @@
...
-        "types": "./dist/types/oidc-client-ts.d.ts",
-        "default": "./dist/esm/oidc-client-ts.js"
-      },
-      "require": {
-        "types": "./dist/types/oidc-client-ts.d.ts",
-        "default": "./dist/umd/oidc-client-ts.js"
-      }
+      "import": "./dist/esm/oidc-client-ts.js",
+      "require": "./dist/umd/oidc-client-ts.js",
+      "types": "./dist/types/oidc-client-ts.d.ts"
```
Comment (kherock) [inline package.json:17]: Make sure that `"types"` is listed first. TS will evaluate the "import" condition for cases where the the `.d.ts` is stored as a sibling

image

Comment (josundt) [reply package.json]: Done

**Comment by josundt** [comment]
@kherock Any chance you could quickly release a new package once merged to resolve this TypeScript 4.7 issue?

**Comment by pamapa** [comment]
i will release once this MR is approved and merged, hopefully this week

**Comment by pamapa** [comment]
thanks for contributing

**Pull request merged**